attiny402_firmware/src/serial.cpp

```diff
@@ -58,67 +45,67 @@ void handler::write_pressure(uint16_t pressure){
 void handler::write_id(){
     servo_write(
         id,
-        SENSOR_ID_READ_LENGTH,
+        SENSOR_ID_READ_RES_LENGTH,
         SENSOR_ID_READ,
-        eeprom_read_byte((uint8_t*)EEPROM_ID_ADDR)
+        DEFAULT_SENSOR_ID
 	);
   
   	Serial.flush(); //may cause slower write
 }
 
 void handler::handle(){
-    for(auto b : expected_header){          //check if header and id are correct
-        while(!Serial.available());         //wait for bytes 
-        if(Serial.read() != b) return;
-    }  
-    
+    while(!Serial.available());
+    if(Serial.read() != expected_headers[0][0]) return;
+    while(!Serial.available());
+    if(Serial.read() != expected_headers[0][1]) return;
+    while(!Serial.available());
+    uint8_t res = Serial.read();
+    if(res != expected_headers[0][2] && res != expected_headers[1][2]) return;
+
     while(Serial.available() < 2);
     uint8_t length = Serial.read();         //read packet length and cmd id
     uint8_t cmd = Serial.read();
+    uint8_t _id = res;
 
     uint8_t buf[7] = {};                    //array to store packet
 
-    uint8_t sum = id + length + cmd;
-    
+    uint8_t sum = _id + length + cmd;
+
     while(Serial.available() < length - 2); //wait for bytes
 
     for(uint8_t i = 0; i < length - 2; i++){//iterate until the packet has been fully read
         buf[i] = Serial.read();
-        if(i != length - 1) sum += buf[i];  //don't add checksum to calculated checksum
+        if(i != length - 3) sum += buf[i];   //don't add checksum to calculated checksum
     }
 
-    delayMicroseconds(500);
+    delayMicroseconds(100);
 
-    //if(~sum != buf[length - 1]) return;     //match checksum
+    if(((~sum) & 0xFF) != (buf[length - 3] & 0xFF)) return;     //match checksum
```

Comment:
Since both sum and *buf are uint8_t `& 0xFF` shouldn't be doing anything

Reply:
wasn't this the file we checked together a couple of days ago? if it's this then *somehow* we need those `& 0xFF` else it doesn't work

Reply:
Wasn't it for the brackets?

Reply:
we should do more testing on that, I believe it was both brackets and the and

Reply:
it didn't work without the and, I think it was doing an implicit cast to something else while evaluating the ~ operator

Reply:
Apparently it is casted to int and then inverted, it's better to leave them then